### PR TITLE
fix(master) alert truenas HA customers properly

### PR DESCRIPTION
**Pull request by yocalebo**
ticket #39114

There will be dependent PR's in the iXsystems/truenas repo

### CODE REVIEW

**Comment by yocalebo** [comment]
This PR has been tested on TrueNAS HA and seems to be working. I have talked with @themylogin out of band to make sure the majority of this logic makes sense. I'm going to add @william-gr for safe measure.

File: src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+
+async def setup(middleware):
+    if os.path.exists(SENTINEL_PATH):
+        # We want to emit the mail only if the machine truly rebooted
+        if os.path.exists(MIDDLEWARE_STARTED_SENTINEL_PATH):
+            return
+
+        gc = await middleware.call('datastore.config', 'network.globalconfiguration')
+        hostname = f"{gc['gc_hostname']}.{gc['gc_domain']}"
```
[the remark this answers is not in the document]
Comment (william-gr) [reply src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py]: This is not about active/passive, its about head identification, "A" or "B".

Like this hostname will always be of the same machine, no matter who is what.

File: src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+            else:
+                await middleware.call("mail.send", {
+                    "subject": f"{hostname}: Failover event",
+                    "text": textwrap.dedent(f"""\
+                        {hostname} had a failover event.
+                        The system was rebooted because persistent SCSI reservations were lost and/or cleared.
+                        The operating system successfully came back online at {now}.
+                    """),
+                })
```
Comment (amotin) [inline src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py]: Not critical, but I would reorganize this code to not duplicate messages.  Something like `if (exist(A) and (!exist (B) or mtime(A) > mtime(B)) ... elif (exist(B)) ... else ...`

File: src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+async def setup(middleware):
+    if os.path.exists(SENTINEL_PATH):
+        # We want to emit the mail only if the machine truly rebooted
+        if os.path.exists(MIDDLEWARE_STARTED_SENTINEL_PATH):
+            return
+
+        gc = await middleware.call('datastore.config', 'network.globalconfiguration')
+        now = datetime.now().strftime("%c")
+
+        Client() = c
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py]: use `await middleware.call('notifier.is_freenas')` instead. Client() will be blocking and hang event loop.

File: src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        if not c.call('notifier.is_freenas') and c.call('notifier.failover_licensed')
+            try:
+                with open(TRUENAS_HA_FILE, "rb") as controller:
+                    match = re.search(r"(.*):(.*)")
+                if match.group(2) == 'B':
+                    hostname = f"{gc['gc_hostname_b'].{gc['gc_domain']}"
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/unscheduled_reboot_alert.py]: You're not handling node = 'A"

Also, better use: notifier.failover_node call to get B or A

**Pull request merged**